pkg/daemon/containermeta/container_meta_controller.go

```diff
@@ -0,0 +1,306 @@
+	if pod.DeletionTimestamp == nil && len(pod.Status.ContainerStatuses) > 0 {
+		key := pod.Namespace + "/" + pod.Name
+		if duration == 0 {
+			q.Add(key)
+		} else {
+			q.AddAfter(key, duration)
+		}
+	}
+}
+
+func (c *Controller) Run(stop <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting containermeta Controller")
+	for i := 0; i < workers; i++ {
+		go wait.Until(func() {
+			for c.processNextWorkItem() {
+			}
+		}, time.Second, stop)
+	}
+
+	klog.Info("Started containermeta Controller successfully")
+	<-stop
+}
+
+// processNextWorkItem will read a single work item off the workqueue and
+// attempt to process it, by calling the syncHandler.
+func (c *Controller) processNextWorkItem() bool {
+	// pull the next work item from queue.  It should be a key we use to lookup
+	// something in a cache
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.sync(key.(string))
+
+	if err == nil {
+		// No error, tell the queue to stop tracking history
+		c.queue.Forget(key)
+	} else {
+		// requeue the item to work on later
+		c.queue.AddRateLimited(key)
+	}
+
+	return true
+}
+
+func (c *Controller) sync(key string) (retErr error) {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		klog.Warningf("Invalid key: %s", key)
+		return nil
+	}
+
+	pod, err := c.podLister.Pods(namespace).Get(name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		klog.Errorf("Failed to get Pod %s/%s from lister: %v", namespace, name, err)
+		return err
+	} else if pod.DeletionTimestamp != nil || len(pod.Status.ContainerStatuses) == 0 {
+		return nil
+	}
+	resourceVersionExpectation.Observe(pod)
+	if satisfied, duration := resourceVersionExpectation.IsSatisfied(pod); !satisfied {
+		if duration < maxExpectationWaitDuration {
+			return nil
+		}
+		klog.Warningf("Wait for Pod %s/%s resourceVersion expectation over %v", namespace, name, duration)
+		resourceVersionExpectation.Delete(pod)
+	}
+
+	_, kubeRuntime, err := c.getRuntimeForPod(pod)
+	if err != nil {
+		klog.Errorf("Failed to get runtime for Pod %s/%s: %v", namespace, name, err)
+		return nil
+	}
+
+	klog.V(3).Infof("Start syncing for %s/%s", namespace, name)
+	defer func() {
+		if retErr != nil {
+			klog.Errorf("Failed to sync for %s/%s: %v", namespace, name, retErr)
+		} else {
+			klog.V(3).Infof("Finished syncing for %s/%s", namespace, name)
+		}
+	}()
+
+	kubePodStatus, err := kubeRuntime.GetPodStatus(pod.UID, name, namespace)
+	if err != nil {
+		return fmt.Errorf("failed to GetPodStatus: %v", err)
+	}
+
+	containerMetaSet := c.generateContainerMetaSet(pod, kubePodStatus)
+	oldContainerMetaSet, err := appspub.GetRuntimeContainerMetaSet(pod)
+	if err != nil {
+		klog.Warningf("Failed to get old runtime meta from Pod %s/%s: %v", namespace, name, err)
+	} else if reflect.DeepEqual(containerMetaSet, oldContainerMetaSet) {
+		return nil
+	}
+
+	pod = pod.DeepCopy()
```

Comment:
there is no need to copy every bit of pod during a status update since only the meta and status is used by the apiserver. We can save the cpu time for deepcopy pod spec